scenario/scripts/snapshot.py

```diff
@@ -0,0 +1,752 @@
+        self.unit_name = unit_name
+        self.app_name = app_name
+        self.unit_id = int(unit_id)
+        self.normalized = f"{app_name}-{unit_id}"
+
+
+def _try_format(string):
+    try:
+        import black
+
+        try:
+            return black.format_str(string, mode=black.Mode())
+        except black.parsing.InvalidInput as e:
+            logger.error(f"error parsing {string}: {e}")
+            return string
+    except ModuleNotFoundError:
+        logger.warning("install black for formatting")
+        return string
+
+
+def format_state(state):
+    return _try_format(repr(state))
+
+
+def format_test_case(state, charm_type_name=None, event_name=None):
+    ct = charm_type_name or "CHARM_TYPE  # TODO: replace with charm type name"
+    en = event_name or "EVENT_NAME,  # TODO: replace with event name"
+    return _try_format(
+        dedent(
+            f"""
+from scenario.state import *
+from charm import {ct}
+
+def test_case():
+    state = {state}
+    out = state.trigger(
+        {en}
+        {ct}
+        )
+
+"""
+        )
+    )
+
+
+def _juju_run(cmd, model=None) -> Dict[str, Any]:
+    _model = f" -m {model}" if model else ""
+    raw = run(
+        f"""juju {cmd}{_model} --format json""".split(), capture_output=True
+    ).stdout.decode("utf-8")
+    return json.loads(raw)
+
+
+def _juju_ssh(target: Target, cmd, model: Optional[str] = None) -> str:
+    _model = f" -m {model}" if model else ""
+    command = f"""juju ssh{_model} {target.unit_name} {cmd}"""
+    raw = run(command.split(), capture_output=True).stdout.decode("utf-8")
+    return raw
+
+
+def _juju_exec(target: Target, model, cmd) -> str:
+    # action-fail              juju-reboot         payload-unregister  secret-remove
+    # action-get               jujuc               pod-spec-get        secret-revoke
+    # action-log               k8s-raw-get         pod-spec-set        secret-set
+    # action-set               k8s-raw-set         relation-get        state-delete
+    # add-metric               k8s-spec-get        relation-ids        state-get
+    # application-version-set  k8s-spec-set        relation-list       state-set
+    # close-port               leader-get          relation-set        status-get
+    # config-get               leader-set          resource-get        status-set
+    # containeragent           network-get         secret-add          storage-add
+    # credential-get           open-port           secret-get          storage-get
+    # goal-state               opened-ports        secret-grant        storage-list
+    # is-leader                payload-register    secret-ids          unit-get
+    # juju-log                 payload-status-set  secret-info-get
+    _model = f" -m {model}" if model else ""
+    _target = f" -u {target}" if target else ""
+    return run(
+        f"juju exec{_model}{_target} -- {cmd}".split(), capture_output=True
+    ).stdout.decode("utf-8")
```

Comment:
This is just personal preference, but I would not have all the possible juju commands listed here. It adds 13 lines that you really do not need. You could have something like the following docstrings:

```python
"""Execute a juju command.

Notes:
    Visit the Juju documentation to view all possible Juju commands:
    https://juju.is/docs/olm/juju-cli-commands
"""
```
Easy way to make maintaining the command reference someone elses problem :smile:

Also, just because I do not like one-use variables:

```suggestion
def _juju_exec(target: Target, model: str, cmd: str) -> str:
    return run(
        shlex.split(
            (
                f"juju exec{(f' -m {model}' if model else '')}
                f"{(f' -u {target}' if target else '')} -- {cmd}"
            )
        ),
        capture_output=True,
    ).stdout.decode("utf-8")
```
_I am assuming that the type of model and cmd is `str`_

Reply:
and, all these single-use vars you mention, I find them very handy when I'm debugging as they allow me to inspect what the command is that's causing a failure at a glance (or I can print locals()...)